### PR TITLE
Simple weather setup

### PR DESCRIPTION
**Pull request by SimonKaya91**
### Things Done:
Added unit tests for weather.find method within getWeather using sinon stubs.
Removed Karma for testing
Removed Express and replaced with Restify.
Created a static resource for testing weather.find method.

Updated:
Actioned comments.

### Instructions:
N/A

### Comments:
First time using Sinon for testing callbacks check over tests are of good standard and practice.

### CODE REVIEW

File: spec/forecasts/forecast.js [hunk 1]
```diff
@@ -1,9 +1,79 @@
...
+import * as weather from "weather-js";
+import {staticWeather} from "../static-resource/weather.json"; //Mock data
 
-describe("here1>", () => {
+describe("Forecast", () => {
+  let forecast;
 
-  it('will assert true to equal true', () => {
+  describe("getWeather", () => {
+    let expectedErrMsg;
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: call full name expectedErrorMessage?

File: spec/forecasts/forecast.js [hunk 2]
```diff
@@ -1,9 +1,79 @@
...
 
-    expect(true).to.equal(true);
+    beforeEach(() => {
+      forecast = new Forecast();
+      expectedErrMsg = "Values undefined";
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: same as above

File: spec/forecasts/forecast.js [hunk 3]
```diff
@@ -1,9 +1,79 @@
...
+    });
+
+    afterEach(() => {
+      forecast = null;
+    });
+
+    describe("with falsy params", () => {
+      it("Return early when either params undefined", () => {
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: when both params undefined*

File: spec/forecasts/forecast.js [hunk 4]
```diff
@@ -1,9 +1,79 @@
...
+
+      it("Return early when params are NaN", () => {
+        forecast.getWeather(NaN, NaN, (err, result) => {
+          expect(err).to.equal(expectedErrMsg);
+        });
+      });
+
+    });
+
+    describe("Weather.find callback result", () => {
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: when weather.find is called

File: spec/forecasts/forecast.js [hunk 5]
```diff
@@ -1,9 +1,79 @@
...
+        stub = sinon.stub(weather, "find"); //Stub any callout to this method
+        expectedErrMsg = "CallbackError";
+      });
+
+      afterEach(() => {
+        forecast = null;
+        stub.restore();
+      });
+
+      it("Tests getWeather returns json object", () => {
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: should return the yielded result

File: spec/forecasts/forecast.js [hunk 6]
```diff
@@ -1,9 +1,79 @@
...
+        stub.yields(null, staticWeather); //Mock test data
+        forecast.getWeather("London, England", "C", (err, result) => {
+          expect(result[0].name).to.equal(staticWeather[0].name);
+          expect(result[0].degreetype).to.equal(staticWeather[0].degreetype);
+        });
+      });
+
+      it("Tests getWeather callback throws an error", () => {
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: should return the yielded error

File: spec/forecasts/forecast.js [hunk 7]
```diff
@@ -1,9 +1,79 @@
...
+        stub.yields(expectedErrMsg);
+        forecast.getWeather("London, England", "C", (err, result) => {
+          expect(err).to.equal(expectedErrMsg);
+        });
+      });
+
+    });
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: Could add test after this to asser that weather.find is called with the right params passed into out function

File: src/forecasts/forecast.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import * as weather from "weather-js";
+
+export class Forecast {
+
+  getWeather(location, degreeType, callback) {
+    //Ensure values exist
```
Comment (reecemillsom) [inline src/forecasts/forecast.js]: remove comment

File: spec/forecasts/forecast.js [hunk 9]
```diff
@@ -1,9 +1,79 @@
...
+
+    describe("Weather.find callback result", () => {
+      let stub;
+
+      beforeEach(() => {
+        forecast = new Forecast();
+        stub = sinon.stub(weather, "find"); //Stub any callout to this method
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: remove comment

File: spec/forecasts/forecast.js [hunk 10]
```diff
@@ -1,9 +1,79 @@
+import {Forecast} from "../../src/forecasts/forecast.js";
+import * as weather from "weather-js";
+import {staticWeather} from "../static-resource/weather.json"; //Mock data
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: remove comment

**Comment by reecemillsom** [comment]
Can we remove test-main.js as part of this also as we are removing the karam config. Not needed anymore.

File: spec/forecasts/forecast.js (hunk 9 above)
Comment (reecemillsom) [reply spec/forecasts/forecast.js]: remove stub = and reference it as weather.find, stub is to vague

File: spec/forecasts/forecast.js [hunk 11]
```diff
@@ -1,9 +1,79 @@
...
+
+      beforeEach(() => {
+        forecast = new Forecast();
+        stub = sinon.stub(weather, "find"); //Stub any callout to this method
+        expectedErrMsg = "CallbackError";
+      });
+
+      afterEach(() => {
+        forecast = null;
+        stub.restore();
```
Comment (reecemillsom) [inline spec/forecasts/forecast.js]: This will become weather.find.restore();

File: spec/forecasts/forecast.js (hunk 5 above)
Comment (reecemillsom) [reply spec/forecasts/forecast.js]: or should callback with the result

File: spec/forecasts/forecast.js (hunk 6 above)
Comment (reecemillsom) [reply spec/forecasts/forecast.js]: or should callback with the error

**Comment by reecemillsom** [comment]
Another thing i'm not sure if it is worth adding is checking that the degreeType is either F or C and if not returning early in our getWeather function. Or we can handle this some way through the UI later on? what do people think?

File: spec/forecasts/forecast.js (hunk 1 above)
Comment (SimonKaya91) [reply spec/forecasts/forecast.js]: Quite a long variable name anyhow was trying to make it shorter. Can change it if you feel really needs to be full name personally think it is descriptive enough though?

Comment (reecemillsom) [reply spec/forecasts/forecast.js]: Good Practice to use full names, don't worry how long the variable name as the more explicit it is the better.

File: spec/forecasts/forecast.js (hunk 7 above)
Comment (SimonKaya91) [reply spec/forecasts/forecast.js]: Not sure what you mean here? We check for false params and use correct ones for the stubs do you mean run another get weather test and check it doesn't return early?

Comment (reecemillsom) [reply spec/forecasts/forecast.js]: Sorry i mean do a new it block, and ensure what we passed in, is used when calling weather.find. so your assert will be checking weather.find it called with

**Comment by SimonKaya91** [comment]
@reecemillsom 
Where is the Karma config file can't see it any more thought we already got rid of it?

For the degree type don't see it worth checking as it's just a preference on what value type the user wants the result in.

**Comment by reecemillsom** [comment]
@SimonKaya91 it should be called test-main in the root of the project,  and my point about checking that is it depends how we handle it in the UI. If we allow them to put it in a input box we need that to be handled to say not valid type, or if we have checkbox for celcius and farenheight then should be fine and wont need anything doing on server

**Comment by SimonKaya91** [comment]
Yeah I don't think we should be handling that server side can just restrict it like you say. 
I have actioned test-main already just thought there was another karma.config file somewhere.

**Comment by RaptorGB** [comment]
Merged, project set up for start now so branch deleted not needed anymore.

**Pull request merged**